src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+(define-map reward-cycle-total-stacked
+    ((reward-cycle uint))
+    ((total-ustx uint))
+)
+
+;; Internal map read by the Stacks node to iterate through the list of
+;; PoX reward addresses on a per-reward-cycle basis.
+(define-map reward-cycle-pox-address-list
+    ((reward-cycle uint) (index uint))
+    (
+        (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+        (total-ustx uint)
+    )
+)
+
+(define-map reward-cycle-pox-address-list-len
+    ((reward-cycle uint))
+    ((len uint))
+)
+
+;; When is a PoX address active?
+;; Used to check of a PoX address is registered or not in a given
+;; reward cycle.
+(define-map pox-addr-reward-cycles
+    ((pox-addr (tuple (version uint) (hashbytes (buff 20)))))
+    (
+        (first-reward-cycle uint)
+        (num-cycles uint)
+    )
+)
+
+;; Binding between the principal whose tokens are stacked, and the principal who
+;; is allowed to set the PoX reward address.  This binding also includes permissions
+;; that the delegator has, as described below.
+(define-map delegates
+    ;; The address of who owns the tokens to be stacked.
+    ((stacker principal))
+    (
+        ;; The address of the delegate, who stacks them.
+        (delegate principal)
+        ;; Number of uSTX this delegate may stack on this stacker's behalf.
+        (amount-ustx uint)
+    )
+)
+
+;; Binding between the delegate, and the amount of STX and PoX addresses it 
+;; is entrusted with.  Updated when the client delegates STX.
+(define-map delegate-control
+    ((delegate principal))
+    (
+        ;; Total number of uSTX locked to this delegate
+        (total-ustx uint)
+        ;; PoX address the delegate must use.
+        (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+        ;; Earliest time at which the delegate can Stack the delegated STX.
+        (burn-block-height-start uint)
+        ;; Beginning of this delegate's tenure
+        (first-reward-cycle uint)
+        ;; How long the uSTX will be locked for, in reward cycles
+        (tenure-cycles uint)
+    )
+)
+
+;; Amount of uSTX that reject PoX, by reward cycle
+(define-map stacking-rejection
+    ((reward-cycle uint))
+    ((amount uint))
+)
+
+;; Who rejected in which reward cycle
+(define-map stacking-rejectors
+    ((stacker principal) (reward-cycle uint))
+    ((amount uint))
+)
+
+;; Getter for stacking-rejectors
+(define-read-only (get-pox-rejection (stacker principal) (reward-cycle uint))
+    (map-get? stacking-rejectors { stacker: stacker, reward-cycle: reward-cycle }))
+
+;; Has PoX been rejected in the given reward cycle?
+(define-read-only (is-pox-active (reward-cycle uint))
+    (let (
+        (reject-votes 
+            (default-to
+                u0
+                (get amount (map-get? stacking-rejection { reward-cycle: reward-cycle }))))
+    )
+    ;; (100 * reject-votes) / total-liquid-ustx < pox-rejection-fraction
+    (< (/ (* u100 reject-votes) total-liquid-ustx) (var-get pox-rejection-fraction)))
```

Comment:
Does this calculation take into account the rounding down when dividing? A safer alternative could be:
```
(< (* u100 reject-votes) 
   (* (var-get pox-rejection-fraction) total-liquid-ustx))
```

Reply:
Sure, I can do this.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/2b384d527347cb14e1e84d2a42fc004a33ab2550